Raygun4UWP/RaygunClient.cs

```diff
@@ -0,0 +1,455 @@
+    /// It is best to call this method within a try/catch block.
+    /// If the application is crashing due to an unhandled exception, use the synchronous methods instead.
+    /// </summary>
+    /// <param name="raygunCrashReport">The RaygunCrashReport to send. This needs its OccurredOn property
+    /// set to a valid DateTime and as much of the Details property as is available.</param>
+    public async Task SendAsync(RaygunCrashReport raygunCrashReport)
+    {
+      await SendOrSaveCrashReport(null, raygunCrashReport);
+    }
+
+    /// <summary>
+    /// Sends a crash report immediately to Raygun for the given <see cref="Exception"/>.
+    /// </summary>
+    /// <param name="exception">The <see cref="Exception"/> to send in the crash report.</param>
+    /// <param name="tags">An optional list of tags to send with the crash report.</param>
+    /// <param name="userCustomData">Optional custom data to send with the crash report.</param>
+    public void Send(Exception exception, IList<string> tags = null, IDictionary userCustomData = null)
+    {
+      StripAndSend(exception, tags, userCustomData);
+    }
+
+    /// <summary>
+    /// Sends a RaygunCrashReport immediately to Raygun.
+    /// </summary>
+    /// <param name="raygunCrashReport">The RaygunCrashReport to send. This needs its OccurredOn property
+    /// set to a valid DateTime and as much of the Details property as is available.</param>
+    public void Send(RaygunCrashReport raygunCrashReport)
+    {
+      SendOrSaveCrashReport(null, raygunCrashReport).Wait(3000);
+    }
+
+    private void Application_UnhandledException(object sender, Windows.UI.Xaml.UnhandledExceptionEventArgs e)
+    {
+      Send(e.Exception);
+    }
+
+    private bool InternetAvailable()
+    {
+      IEnumerable<ConnectionProfile> connections = NetworkInformation.GetConnectionProfiles();
+      var internetProfile = NetworkInformation.GetInternetConnectionProfile();
+
+      bool internetAvailable = connections != null && connections.Any(c =>
+                                 c.GetNetworkConnectivityLevel() == NetworkConnectivityLevel.InternetAccess) ||
+                               (internetProfile != null && internetProfile.GetNetworkConnectivityLevel() == NetworkConnectivityLevel.InternetAccess);
+      return internetAvailable;
+    }
+
+    private async Task SendOrSaveCrashReport(Exception originalException, RaygunCrashReport raygunCrashReport)
+    {
+      if (ValidateApiKey())
+      {
+        bool canSend = OnSendingCrashReport(originalException, raygunCrashReport);
+        if (canSend)
+        {
+          try
+          {
+            JsonSerializerSettings settings = new JsonSerializerSettings
+            {
+              NullValueHandling = NullValueHandling.Ignore,
+              Formatting = Formatting.None
+            };
+
+            string payload = JsonConvert.SerializeObject(raygunCrashReport, settings);
+
+            if (InternetAvailable())

```

Comment:
Rather than checking for Internet connectivity every time we send a crash report, we should register for network status changes, see https://docs.microsoft.com/en-us/uwp/api/windows.networking.connectivity.networkinformation.networkstatuschanged
We could also extract this out to a helper that could store a connected/not connected flag so most calls to check connectivity just return the current flag value.  The implementation of `InternetAvailable` could be moved to the helper and still used if we encounter an error trying to send crash reports, e.g. in exception handler for `SendCrashReport`

Reply:
Great idea, thanks for pointing out this event. I've added this to the new HttpService.